### DISCUSSION TITLE
[Feature Request] type check variable assignments when executing under pytest

### DISCUSSION THREAD

**Discussion by frrad** [beartype/beartype#419] [Q&A]
Suppose you have some test:

```
my_package/cursed_test.py:

def test_curses():
    x: int = "nope"

    assert True
```

in some package with 

```
my_package/__init__.py:

from beartype.claw import beartype_this_package

beartype_this_package()
```

If you are me, then you might spend an embarrassingly long time wondering why this happens

```
pytest ./
============================================================== test session starts ==============================================================
platform linux -- Python 3.8.19, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/i_am_very_dumb/test
collected 1 item                                                                                                                                

my_package/cursed_test.py .                                                                                                               [100%]

=============================================================== 1 passed in 0.12s ===============================================================
```

before realizing that this

```
cat main.py 
from my_package.cursed_test import test_curses

test_curses()
```

throws an error `beartype.roar.BeartypeDoorHintViolation: Callable my_package.cursed_test.test_curses() local variable "x" value 'nope' violates type hint <class 'int'>, as str 'nope' not instance of int.` Like I expected to happen in the first place and deducing (hopefully correctly, what do I know) that pytest must be executing the tests in some manner that doesn't involve importing the package, thereby sidestepping the `beartype_this_package` hook.

I would love for there to be a way to run beartype during test execution with pytest. Note that, even if I'm willing to decorate all my tests with `@beartype`, this will still not catch my variable assignment problems.

1. Potentially dumb question: Is there a way to get the `die_if_unbearable` secretly everywhere behavior other than with the package level hook that I've missed? If there were something similar to the `@beartype` decorator I could at least manually decorate a bunch of things in the test file.
2. Feature request! Some kind of magic that just makes it work the way I thought it would originally ;)
3. Do you think this corner case is worth a mention in the FAQ? Maybe I'm just uniquely dense, but I did struggle on this for a while. Part of my motivation for writing this up is so that this issue will at least be some form for documentation for future confused people.

**Comment by leycec** [comment]
**Oh, ho, ho.** BizarroWorld™ packaging directory structure, huh? Yeah. That's... that's your problem. :rofl: 

I hope you don't mind, but I've migrated this into a Q&A discussion format for readability and clarity. Please don't mind! You have totally valid concerns, but they probably intersect better with the forum format. Okay. So...

As you've correctly surmised:

> ...`pytest` must be executing the tests in some manner that doesn't involve importing the package, thereby sidestepping the `beartype_this_package` hook.

**That's it.** That's totally it. You've nailed it in one. This might be surprising. Like @beartype, `pytest` itself makes heavy use of import hooks to do its perfidious magic. `assert` rewriting, test collection, and more? Yeah. That's import hooks all the way down. To preserve compatibility with `pytest` import hooks, @beartype import hooks actually go *waaaaaay* out of their way to interoperate cleanly with what `pytest` wants to do.

What `pytest` mostly wants to do is for wonderful packagers such as yourself to **cleanly separate tests from non-tests.** Instead, you've interleaved your tests (e.g., the `my_package.cursed_test` submodule) with your non-test code (e.g., the `my_package.__init__` submodule). Technically, there *probably* are really convoluted workarounds to achieve this. Pragmatically, it's best to never even go there. Interleaving tests and non-tests together in the same `my_package/` directory structure is a recipe for cumbersome migraine-inducing discussions like this. In the immortal words of 80's pop-punk trash outfit *Offspring*:

> You gotta keep 'em separated.

`pytest` really wants you to structure your package files like this:

```
my_package/           # <-- *PYTEST HATES TESTS IN YOUR MAIN PACKAGE*
    __init__.py       # <-- beartype_this_package() for the win!
    curse_it.py       # <-- no tests in here -- only logical business logic

my_package_test/      # <-- *PYTEST LOVES TESTS OVER HERE*
    test_curse_it.py  # <-- all tests over here
```

**So.** What do those two new `my_package.curse_it` and `my_package_test.test_curse_it` submodules look like? Behold! A worky test suite that curses like a drunken sailor while otherwise behaving sanely:

* In the `my_package.curse_it` submodule:

  ```python
  def curse_like_a_sailor():
      x: int = "nope"
      assert True
  ```

* In the `my_package_test.test_curse_it` submodule:

  ```python
  def test_curse_like_a_sailor():
      from beartype.roar import BeartypeDoorHintViolation
      from my_package.curse_it import curse_like_a_sailor
      from pytest import raises

      with raises(BeartypeDoorHintViolation):
          curse_like_a_sailor()
  ```

**Totally works.** Right?

```bash
$ ls
my_package/        my_package_test/
$ pytest
================================================= test session starts =================================================
platform linux -- Python 3.8.19, pytest-7.2.1, pluggy-1.0.0
rootdir: /home/leycec/tmp/
collected 1 item                                                                                                      

my_package_test/test_curse_it.py .                                                                                    [100%]

================================================== 1 passed in 0.31s ==================================================
```

@beartype and `pytest` :heart_hands: each other. @beartype was built from the ground up to play nicely with `pytest`. But @beartype and `pytest` both have hard and fast rules concerning this integration, which mostly just reduces to: <sup>*...waitforit*</sup>

> You gotta keep 'em separated.

![](https://media1.tenor.com/m/_ohy0V2mDfYAAAAd/captain-and-tennille-love-will-keep-us-together.gif)
<sup>`pytest` (diva in front) + @beartype (guy in back): a match made in GitHub</sup>

**Comment by frrad** [comment]
Thank you so much for responding so quickly and comprehensively! I have fixed my crazy directory structure in my toy example, and I believe my problem still remains. Namely, I don't get any errors and don't know how to set things in such a way that I would get errors for incorrect type hints in variable assignments in test files.

Alternatively, I'm still making some crazy mistake... Here's my setup.

```
my_package
├── implementation.py
└── __init__.py
my_package_test
├── __init__.py
└── test_cursed_things.py
```

`my_package/implementation.py`:
```
def totally_valid_function(x: int) -> int:
    return x
```
`my_package/__init__.py`
```
from beartype.claw import beartype_this_package

beartype_this_package()
```
`my_package_test/__init__.py`
```
from beartype.claw import beartype_this_package

beartype_this_package()
```
`my_package_test/test_cursed_things.py`
```
from beartype import beartype
from my_package.implementation import totally_valid_function


@beartype
def test_curses():
    x: int = "nope"

    assert True
```

running `pytest ./`
```
========================= test session starts =========================
platform linux -- Python 3.8.19, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/how_python_wurk_idk/Projects/test
collected 1 item                                                      

my_package_test/test_cursed_things.py .                         [100%]

========================== 1 passed in 0.14s ==========================
```

**Comment by leycec** [comment]
**Oh, ho, ho.** Madness intensifies. So you want to apply `beartype.claw` import hooks to both your main codebase *and* your test suite, huh? That's... a *fascinating* idea. You're the first @beartype user brave enough to sacrifice your sanity for that. It's not easy being brave. Applying `beartype.claw` import hooks to your main codebase is, of course, trivial. It's applying `beartype.claw` import hooks to your *test suite* that's the issue. That might be infeasible. `</gulp>`

How familiar are you with `pytest` hooks, plugins, and `conftest.py` scripts? My strong suspicion is that you'll need to "go there." `pytest` appears to be circumventing standard packaging semantics and `importlib` machinery in your test suite. Kinda sounds like a horrifying can of worms poised to explode all over your face, honestly.

Was there a pressing reason that you needed `beartype_this_package()` to apply to your test suite? Are your authoring your own `pytest` plugin, for example, that you needed to runtime type-check with @beartype? May the Gods help us, for `pytest` has surely abandoned us in our time of testing need. :crying_cat_face:

**Comment by frrad** [comment]
This is likely a meaningless distinction, but I'm not trying to "apply `beartype.claw` import hooks" per se. I just want type enforcement for variables in tests. I'm okay with any means of accomplishing this short of manually rewriting to add `die_if_unbearable` on every other line. Maybe the import hook is the only way of doing this but just throwing that out there.

I am not at all familiar with the inner workings of pytest. It's all just copy-paste and magic to me... I'm probably just going to drop this one. It's not too critical and if it sounds like a can of worms to you I'm sure I would be in for a bad time trying to get it to work.

A little more about my use case in case you're interested:

I recently started using `jaxtyping` to typecheck some of my `torch.Tensor`s. It's very nice for readability to be able to annotate their dimension. Then, I forgot to update one of my annotations as I was editing a test, test still passed and we found ourselves here... FWIW this doesn't feel like a totally crazy thing to want to be able to do. Type hints are 10x nicer when you can trust them. I will likely stop typing the tensors in my tests beyond `torch.tensor` because I'd prefer to avoid having potentially misleading hints in the code. This will make the tests slightly harder to read than they would otherwise be. Far from the end of the world, but not ideal either.

Anyways, Thanks again for your assistance!! Have a beartastic day :)

**Reply by leycec** [reply]
> I just want type enforcement for variables in tests.

**Gotcha!** "Totally reasonable request" is what I would be proudly saying right now if `pytest` were a totally reasonable test framework. I love `pytest`. We all love `pytest`. Nonetheless, `pytest` is the diametric opposite of a totally reasonable test framework. `pytest` is balls-to-the-walls black magic from top to bottom. I question whether `pytest` devs themselves even grok how `pytest` internally works anymore. In response, `pytest` devs are now burning down the @beartype issue tracker.

> I'm okay with any means of accomplishing this short of manually rewriting to add `die_if_unbearable` on every other line.

If I admitted that you have to add `die_if_unbearable` on every other line, how displeased would you be with @beartype on an inclusive scale from `"this is fine"` to `"i will gut you with a trout fishbone the cat dragged in, beartype"`?

> I am not at all familiar with the inner workings of `pytest`.

The inner workings of `pytest` is a GitHub reenactment of **Doom 2: Hell on Earth.**

> ...I'm sure I would be in for a bad time trying to get it to work.

**This.** So much this. Because I love my users, I am trying to spare you painful disappointment coupled with stress-induced kidney disease.

> I recently started using `jaxtyping` to typecheck some of my `torch.Tensor`s.

:partying_face: 

> Then, I forgot to update one of my annotations as I was editing a test, test still passed and we found ourselves here...

That... that's horrible. I concur that the Python QA stack – @beartype, `jaxtyping`, `pytest`, the whole bloody shebang – is failing you here.

> FWIW this doesn't feel like a totally crazy thing to want to be able to do.

No, I totally agree. It's fascinating that you're the first @beartype to want this, because this *is* genuinely useful needful. It's just that `pytest` has locked up @beartype and `jaxtyping` in the closet while running around wearing a balaclava. There's not much @beartype or `jaxtyping` can do about `pytest` being kinda cray-cray, right? `pytest` answers to no one.

> Type hints are 10x nicer when you can trust them.

**Totally.** Kinda the whole point. :laughing: 

> I will likely stop typing the tensors in my tests beyond `torch.tensor`...

**Totally.** Alternately, you could continue typing tensors with well-documented caveats that those types are currently unenforced. Doing so:

* Leaves the door open for **forwards compatibility** with some utopian new version of @beartype, `pytest`, and/or `jaxtyping` that fully supports import hook-based type-checking of `pytest` tests.
* Makes those tensors **self-documenting**. The utility of type hints isn't simply type-checking; it's readability and clarity, too. By design, type hints are the most concise and hopefully accurate description of the internal structure of an object. Even if type hints are currently unenforceable in this context, they still improve intelligibility.

Still, this feature request is sadness. I'm *sooooooooo* sorry that nobody knows how to make @beartype, `pytest`, and `jaxtyping` behave reasonably here. Hopefully, somebody will save us all in a year or two by reverse-engineering `pytest` + @beartype + `jaxtyping` import hook integration. I pity that somebody even as I praise them. :smile_cat:
